### DISCUSSION TITLE
Godot physics and lighting should be able to handle scenes on the centimeter scale

### DISCUSSION THREAD

**Discussion by blackears** [godotengine/godot-proposals#4714] [3D]
This stems from a game I was making where a small ball rolls around in a maze using just the force of gravity and the tilt of the box.  At first, the behavior seemed unrealistic because the ball was moving much too slowly.  Then I realized that this was due to the fact the scale was wrong - my ball was 1m in diameter and was moving correctly for a ball that size.  But since both a large ball and a small one will fall at exactly the same rate due to gravity, a 1m ball viewed from 10 m away will roll much differently than a 1 cm ball viewed from 10 cm away.

So I shrunk my size down to be 1/10 what it originally was, but then started running into new problems.  The physics engine has glitches for 0.1m radius spheres and seems to break down even worse for even smaller objects that are 0.01m.  Cast shadows are also a problem - while my directional light casts shadows correctly for 1m tall objects, the shadows are mostly missing for 0.1m tall objects.

I've already had some discussions about this on the Godot forums:

https://godotforums.org/d/30223-how-to-make-a-rigidbody-roll-more-quickly/8
https://godotforums.org/d/30238-shadows-arent-cast-properly-from-small-objects/4

I feel trying to create games that work on centimeter scales is something that many would be interested in, and in my case gravity just doesn't accelerate objects properly if I'm limited to 1m objects.  If it's a question of resolution, it would be nice if you could provide a hint to the physics and lighting engines that your entire scene fits in a 10m cubed region.

**Comment by Calinou** [comment]
Godot can already support such small scales. If you resize everything to be 10 times smaller, you need to adjust your collision safe margins to be 10 times smaller than they would otherwise be.

For shadows, you need to reduce Shadow Max Distance on the DirectionalLight3D node by a 10× factor and also decrease bias by a 10× factor (but not normal bias, if I'm not mistaken).

That said, there is absolutely no problem with using a much greater gravity scale than the default. If you can't use meters as units, this won't cause any problems in practice as long as you take this into account within your game logic.

> If it's a question of resolution, it would be nice if you could provide a hint to the physics and lighting engines that your entire scene fits in a 10m cubed region.

Floating-point precision is always relative to the scale you're working in, so that wouldn't be useful. If you make everything 100× larger but everything moves 100× faster, the amount of effective precision you're getting is pretty much the same.

**Comment by wareya** [comment]
Increasing the physics framerate to something like 120 or 240 will help as well, but be aware of https://github.com/godotengine/godot-proposals/issues/1893

**Comment by KaijuKoder2** [comment]
Using 4.3 and confirming that Godot can't still handle shadows on small scenes.

On a scene laid out on a 1-meter tile, the shadows are very poor. You have to manipulate the DirectionalLight3D controls to get shadows to appear at all, and even when you do, they're so bad you wouldn't want to use them anyway.

Scale the entire scene by 10, and the shadows are better, but still not great.

Scale the entire scene by 100, and the shadows are quite good.

A few commenters here and on the godotforums eg https://godotforums.org/d/30238-shadows-arent-cast-properly-from-small-objects/4 have said very few people would want to do a small scene, but obviously some people do!!! In my case the only reason to scale up the scene is to get the shadows to work. It increases the complexity of the rest of the program having to factor in the scale change, all for the sake of the shadow algorithm. There's no other reason.

Also it really shouldn't matter: If your camera's view area is small enough, the shadow calculation sizes should be adjusted accordingly. There must be something in the Godot shadow algorithm which is hardcoded for a fairly large scene size.

**Reply by tetrapod00** [reply]
Did you already try changing your camera's `Far` value? 
![Godot_v4 3-stable_win64_ov0XQbGlJy](https://github.com/user-attachments/assets/55b6df1c-403c-4b43-b2ae-c0f21b15af5e)
![Godot_v4 3-stable_win64_aPNnMbSBW3](https://github.com/user-attachments/assets/30e00f0c-35df-44c2-9eda-1e73977aa3c9)

**Reply by KaijuKoder2** [reply]
Thanks for your reply.

I did try a wide range of Far values: There was a very small difference (compare top and bottom at 1X scale), but scaling up by 100X made a huge difference in shadow quality.

![godot shadows problems](https://github.com/user-attachments/assets/954d7d1e-5286-4b8d-a657-a44f5b2c3fdc)

**Reply by tetrapod00** [reply]
It seems that the blur setting on the directional light is at least one setting that is dependent on scale. I can get decently sharp shadows on a 1m plane, with spheres with a radius of .005m. The only settings that matter in my case are a short camera `near` and `far` distance, and a very small blur setting on the light (but not 0).
![Godot_v4 3-stable_win64_5mkfRveDwA](https://github.com/user-attachments/assets/9a0064fa-8299-4db6-8f30-f43035fde7ae)
![Godot_v4 3-stable_win64_mxbLtRnFnU](https://github.com/user-attachments/assets/27e4db21-9b6e-4fc2-b5ec-9c1fa22b0e98)
![Godot_v4 3-stable_win64_gUEpQWYG3O](https://github.com/user-attachments/assets/ee803d1a-3eeb-4870-b9f8-5596230fda8c)

**Comment by KaijuKoder2** [comment]
>  I can get decently sharp shadows on a 1m plane, with spheres with a radius of .005m

The tile is 1m and the trees about the same size as your spheres, but since their shadows are cast by from texture billboards they'd obviously be much thinner, and perhaps so blurry they're hard to discern. No amount of tweaking the Light Shadow settings (could sharpen those shadows???) to get anything close to the 100X scale quality. I did try a 10X scale version too and consistent with that, found that although sharper, it was still a bit too blurry. 

So, yes, Think you are right!
> It seems that the blur setting on the directional light is at least one setting that is dependent on scale.

**Reply by Calinou** [reply]
If your scale becomes much smaller, you also need to decrease **Directional Shadow > Max Distance** accordingly (this way, you don't have to adjust the camera's **Far** property).

You'll also want to decrease **Shadow > Bias** proportionally to the scale, and perhaps **Shadow > Normal Bias** as well.

**Shadow > Blur** is an arbitrary factor, but I don't know if it's intended to be in world-space units or not (in which case it's expected that you have to change it according to scale). However, changing its scaling will break visuals in existing projects, which we should avoid.

**Reply by KaijuKoder2** [reply]
Hey thanks @Calinou and @tetrapod00! I didn't even realize there was a **Directional Shadow**. I thought everything was under the **Shadow** panel so stopped looking. D'oh! Anyhow, setting Max Distance to 1 gave decent results. I did have to set **Shadow > Blur** as low as I could set it 0.001. Setting it higher or lower caused the shadows to disappear, so I guess it does change according to scale.  Perhaps the shadows are a bit blockier than the 100X version(?) not not(!), and it's still pretty good regardless! A forest looks much better with shadows. Thanks!

![Better with Scale 1 Directional Shadow Max Distance 1 Blur 0p001](https://github.com/user-attachments/assets/51bd7947-28ee-45fa-9395-75b6d0e748e7)

**Reply by tetrapod00** [reply]
>  I did have to set Shadow > Blur as low as I could set it 0.001. 

FYI every inspector value is clamped at the inspector UI level - if you override the value from code it will work in most cases, unless the property itself does validation in the setter. So you may be able to set this value lower, within the limits of floating point precision.

**Reply by JustASquid** [reply]
@Calinou Is there any similar setting/workaround for positional shadows? I'm having trouble getting them to look decent in a small scene.

**Reply by Calinou** [reply]
Positional shadow quality largely depends on its range (and for spot lights, also its angle - lower is better). If its range is very large relative to the screen size, shadow quality won't be great unless you increase positional shadow size in the Project Settings.

If you notice shadow peter-panning, you can still decrease bias or normal bias as usual, but be wary of decreasing it too much to avoid introducing shadow acne.
